### ISSUE TITLE
Make text context / per document features available to the pipeline processing steps

### ISSUE THREAD

**Issue by johann-petrak** [explosion/spaCy#4160]
Currently, a pipeline is a function that maps a text to a (processed document) where the tokenizer (which is not explicitly part of the pipeline) first creates the document and where the processing steps of the pipeline then update that document. 

Often, NLP processing requires or may use additional information, not just text: meta information about each document like author, language, topic, classification labels etc. 
Both the tokenizer and the processing steps in the pipeline may want to access those additional features on a document by document basis. 

There is no clean and generic way at the moment to achieve this. The pipe `as_tuples` parameter can be used to pass on input contexts to the consumer AFTER the document has been processed, but it is not possible to access those contexts from either the tokenizer or the processing steps in a pipeline. 

The maybe easiest way to do this would be to add a `context=None` keyword parameter to the tokenizer and processing step interfaces and to add a `use_context=False` keyword parameter to the `pipe` method. If `use_context=True` and `as_tuples` is True as well, the per tuple context would get passed to each `__call__` or `pipe` that is invoked by the top pipe invocation. 

That way the using existing tokenizers and processing steps would not break when `as_tuples=True` is used, but if `use_context=True` is specified as well, all the steps involved would have to already accept the new `context=..` parameter. 

Each processing step could make use of the context however they see fit (eg just storing a feature with a document) or just ignore it. 

Conecptually a pipeline would then be a function that maps a tuple (text, context) or (text, c1, c2, c3) to a document. (if a tuple of the form (text, c1, c2,..) is processed then the list [c1, c2, ..] could be passed on as the context)

**Comment by honnibal** [comment]
Yes this is definitely a problem at the moment, and we've been puzzling over ways to solve it. Broadly we can categorize the possible solutions into:

1. Add optional arguments to `__call__` and `pipe()`.
2. Introduce additional methods, e.g. `component.call_with_context` and `pipe_with_context`
3. Set the context into the `Doc`, using extension attributes.
4. Components take either `Doc` objects or `(doc, context)` tuples as their positional argument input.
5. Set the context as state, either on the components or somewhere they can retrieve it, and keep the current APIs.

If you had to write a component today that used extension attributes without subclassing the `Language` class or making any other changes, you'd probably be forced to do something like 5 (set state somewhere). It's definitely a bad design, but it does work with the minimum changes to anything else. So if we need a while to think about a better solution, we could publish a recipe showing an example of this workflow.

I think 4 (complicating the positional argument) is pretty much strictly worse than 1, so we don't need to talk about it further.

I've leaned towards 3 when thinking about this previously, but I now think it's not a good approach. Amongst other drawbacks, it really locks the core library out of the functionality, since we shouldn't be setting extension attributes within spaCy. It's also pretty weird to force people to unset the data later if they never intended it to get carried along with the doc.

So to me it's between 1 and 2. I sort of regret making the `as_tuples` mechanism a keyword of `Language.pipe()` rather than another method, so I'm worried we would be repeating the same mistake again by adding another keyword. On the other hand, maybe it introduces an inconsistency if this is a new method, while previously we've done this with a keyword.

Another consideration is whether we should make the context a dict that can be addressed by the different component names. I would say this is probably a good idea: otherwise there will be no way to specify context for some pipeline components but not others. You might have two pipeline components that use some generic name, and need to provide different values for the different components.

**Comment by johann-petrak** [comment]
Thanks, interesting analysis. I would also definitely lean towards 1 and 2, but even more towards 1 :)

Ideally, the `as_tuples` parameter would already pass on the context to call and pipe, but implementing this now would mean that other people's components would have to get updated to accept the new context= (or whatever) keyword parameter. This is why I suggested the additional `use_context` parameter to the top pipe method. I would like just to have the semantics of `as_tuples`  to get extended better, but this is a decision about backwards compatibility. 

I think changing things so that a keyword parameter with a default value None is added to call and pipe is minimal impact and looks cleaner than having new methods added. Especially since the `__call__` method is special and we would lose the ability to do `nlp(text, context=something)` and replace it with the much uglier and much less obvious `nlp.call_with_context(text, context)`. 

I am not too fond of distinguishing by component name what context to pass. A context is some opaque object and users are free to use a datastructure that has different data mapped from component names, but I would not force this as in many cases, the context will actually get used only in one component anyway or each component that uses it would know which part to use. 

My suggestion is to simply pass on whatever remains after the first element of a tuple: if there is just one, pass that, otherwise pass the list of what remains. This could then also easily be used to process a list instead of a tuple where by convention, again the text is the first element.

My own current workaround actually is a slight modification of the toplevel pipe which first thing assigns the context to a document attribute, but in general that would be a bad approach because it is easy to forget removing all that data from the document if not needed (any more) and then eat up memory or produce massive serialisations of the documents.

**Comment by honnibal** [comment]
I agree that it's between 1) and 2), and I probably also favour 1).

I do think we need to think about passing arguments to the different pipeline components though, or we'll end up with problems. It's normal that different components will want different contextual features, and components should also have the ability to check for unexpected or invalid contextual features.

Even in the worst-case, where you have a whole pipeline that needs the same context object, the code would look like:

```
doc = nlp(text, context={name: context for name in nlp.pipe_names})
```

Which I think is not so bad. Much more often you will have something like `doc = nlp(text, context={"my_thing": context})`, which seems fine too.

**Comment by johann-petrak** [comment]
As I said, my personal opinion is that I do not really see an advantage, only a limitation with doing this since a user would always be free to pass such a dictionary anyway if necessary by her specific pipeline and each component that knows its name could then pick what it needs, but in general, a context would be just some application specific opaque data structure which is passed to all components and could get ignored. The overhead of creating that additional dictionary you suggest is probably not big even if we process millions of items (in comparison to what else happens in a Spacy pipeline), but it looks unnecessary to me. 
Also, I think only actual processing steps that are in the pipeline have names, but the initial tokenizer / make_doc step does not, but may also want to know the context in some applications. 

Unlike configuration data which is specific to the processing step, I see the context conceptually as something that is essentially equivalent to the text/document which is also passed to all steps equally. 

However it is not terribly important from my POV either, as long as the context can be passed to all components AND the initial tokenizer / document creator, having it to put into a dictionary is fine too.

**Comment by honnibal** [comment]
The problem is really compatibility between components. It's really a namespacing problem, like the one you were talking about in relation to the extension attributes.

Two pipeline component might expect the context to have something like `"location"`, and might expect that to be a country or origin. A different component might optionally support a "location" key in the context too -- but might be expecting something different. Once this is the case, the two pipeline components won't be compatible in the same pipeline, because I'll have no way of providing the location feature to one component without giving an incorrect feature to the other.

It's also very reasonable to want to allow components to validate their contexts. If I have to write my component to expect arbitrary context, I can't alert the user if the context contains unexpected/unsupported items. That means if the user provides the context incorrectly, the component silently ignores those features, and the user will likely just train again and see no effect from the features. That's a fairly annoying problem to debug.

**Comment by johann-petrak** [comment]
I am not feeling strongly either way, and maybe you envision different use cases. 
Just saying that your examples can all be solved on a per application basis by still wrapping the data into a dictionary that provides different chunks of context to different processing stages if necessary. Depending on the use case, component 1 may need chunks 1,2, and 3, component 2 may need chunk 4, component 3 may need chunk 1, 4 and 5 etc. so I believe the way how parts of the context need to get used by each step in the pipeline could potentially be highly application specific anyways. So with my proposal it would still be possible to do something like this or more, if necessary, and the way needed by the applicaton, while with your proposal it would always have to be done in the way the Spacy API requires. 
Let me say though I am happy with either (since as you pointed out, one could still build a dict where the same context gets passed to all components) as long as a context can be passed somehow :)

**Comment by adrianeboyd** [comment]
We're going back through older issues given the upcoming release of spacy v3.

I think we're back to leaning toward option 3, since you can now train directly from `Doc` objects with the new `Example` class and your custom pipeline components can access the custom extensions on both the reference and predicted doc to train / predict / evaluate. We still wouldn't have any support for this in the core pipeline components, however it should be easier to write and extend your own, see: https://nightly.spacy.io/usage/training#config-custom

The custom extensions don't solve the problems around namespacing or unintentionally bloating what's stored by a `Doc`, though.

**Comment by narayanacharya6** [comment]
Sorry to start this conversation again but I just wanted to understand the downsides of adding an optional argument to the `__call__` and `pipe()` methods as suggested in option 1. I feel a simple dictionary (default `None`) for `context` would do the job here. Is there a design consideration I am missing here? 

I have been working around this by adding extensions to the `Doc` object but they keep adding on and when I need to serialize it I need to remove the ones I don't need just to make sure I don't add unnecessary extensions to the serialized dump. Wrapping the responsibility of using certain extra meta-information used by a pipe within the pipe itself feels like the right way to organize code.

**Comment by polm** [comment]
I believe this is resolved by #9069, which allows passing Docs directly to pipelines, skipping the tokenizer and preserving any extra data on the doc.
